ruby/ql/lib/codeql/ruby/ast/Module.qll

```diff
@@ -63,6 +78,99 @@ class Module extends TModule {
           loc.getStartColumn()
       )
   }
+
+  /** Gets a constant or `self` access that refers to this module. */
+  private Expr getAnImmediateReferenceBase() {
+    resolveConstantReadAccess(result) = this
+    or
+    result.(SelfVariableAccess).getVariable() = this.getADeclaration().getModuleSelfVariable()
+  }
+
+  /** Gets a singleton class that augments this module object. */
+  SingletonClass getASingletonClass() { result.getValue() = this.getAnImmediateReferenceBase() }
+
+  /**
+   * Gets a singleton method on this module, either declared as a singleton method
+   * or an instance method on a singleton class.
+   *
+   * Does not take inheritance into account.
+   */
+  MethodBase getAnOwnSingletonMethod() {
+    result.(SingletonMethod).getObject() = this.getAnImmediateReferenceBase()
+    or
+    result = this.getASingletonClass().getAMethod().(Method)
+  }
+
+  /**
+   * Gets an instance method named `name` declared in this module.
+   *
+   * Does not take inheritance into account.
+   */
+  Method getOwnInstanceMethod(string name) { result = this.getADeclaration().getMethod(name) }
+
+  /**
+   * Gets an instance method declared in this module.
+   *
+   * Does not take inheritance into account.
+   */
+  Method getAnOwnInstanceMethod() { result = this.getADeclaration().getMethod(_) }
+
+  /**
+   * Gets the instance method named `name` available in this module, including methods inherited
+   * from ancestors.
+   */
+  Method getInstanceMethod(string name) { result = lookupMethod(this, name) }
+
+  /**
+   * Gets an instance method available in this module, including methods inherited
+   * from ancestors.
+   */
+  Method getAnInstanceMethod() { result = lookupMethod(this, _) }
+
+  /** Gets a constant or `self` access that refers to this module. */
+  Expr getAnImmediateReference() {
+    result = this.getAnImmediateReferenceBase()
+    or
+    result.(SelfVariableAccess).getVariable().getDeclaringScope() = this.getAnOwnSingletonMethod()
+  }
+
+  pragma[nomagic]
+  private string getEnclosingModuleName() {
+    exists(string qname |
+      qname = this.getQualifiedName() and
+      result = qname.regexpReplaceAll("::[^:]*$", "") and
+      qname != result
+    )
+  }
+
+  pragma[nomagic]
+  private string getOwnModuleName() {
+    result = this.getQualifiedName().regexpReplaceAll("^.*::", "")
+  }
+
+  /**
+   * Gets the enclosing module, as it appears in the qualified name of this module.
+   *
+   * For example, the canonical enclosing module of `A::B` is `A`, and `A` itself has no canonical enclosing module.
+   */
+  pragma[nomagic]
+  Module getCanonicalEnclosingModule() { result.getQualifiedName() = this.getEnclosingModuleName() }
```

Comment:
I'm not sure I understand what a "canonical" enclosing module means. Could you give an example of a non-canonical enclosing module?

Reply:
A module can aliased, i.e. reachable via more than one chain of constants. One of these chains of constants is considered to be the "canonical" name, the one used in the `TResolved` constructor, and the rest are just aliases. This predicate gets the enclosing module _as it appears in the canonical name_.

For the module `A::X` below, the module `A` would be its canonical enclosing module.
```rb
module A
end
module B
  module A::X
  end
end
```

Reply:
The problem is that the "canonical" name in `TResolved`  is rather arbitrary and does not always reflect what the Ruby interpreter would report as the "name" . So it may be a bit misleading. In the interpreter a module's name appears to be determined by the name of the first constant to which it is assigned.

Reply:
Is there another name we could use for this concept that you'd find more intuitive?

Reply:
That is a very good question. I couldn't find a name that is used in the Ruby documentation or implementation for the "parent" or "container" of a module/class.

There does not seem to be any reflection method to get that information in the standard library (apart from parsing the value of `Module.name`). However, there is a method named `module_parent` in Rails: https://apidock.com/rails/v6.0.0/Module/module_parent

So perhaps that is the best name for this concept?

Reply:
Pushed a commit to rename:
- `getCanonicalEnclosingModule` -> `getParentModule`
- `getCanonicalNestedModule` -> `getNestedModule`